File: 3519.count-numbers-with-non-decreasing-digits.java
#
# @lc app=leetcode id=3519 lang=java
#
# [3519] Count Numbers with Non-Decreasing Digits
#

# @lc code=start
class Solution {
    private static final int MOD = 1000000007;
    private static final int MAXL = 355;
    private static final int MAXD = 12;

    private long[][][][] dp;
    private int[][][][] visits;
    private int vers;
    private int basee;
    private java.util.List<java.lang.Integer> ddigits;
    private int leng;

    public int countNumbers(String l, String r, int b) {
        dp = new long[MAXL][MAXD][2][2];
        visits = new int[MAXL][MAXD][2][2];
        vers = 0;
        long cntR = solve(r, b);
        String lm1 = decrement(l);
        long cntL = solve(lm1, b);
        return (int) ((cntR - cntL + MOD) % MOD);
    }

    private long solve(String s, int b) {
        if ("0".equals(s)) {
            return 0L;
        }
        java.math.BigInteger num = new java.math.BigInteger(s);
        ddigits = getBaseBDigits(num, b);
        leng = ddigits.size();
        basee = b;
        vers++;
        return dfs(0, 0, 1, 1);
    }

    private long dfs(int pos, int pre, int tig, int lzero) {
        if (pos == leng) {
            return lzero == 0 ? 1L : 0L;
        }
        if (visits[pos][pre][tig][lzero] == vers) {
            return dp[pos][pre][tig][lzero];
        }
        visits[pos][pre][tig][lzero] = vers;
        dp[pos][pre][tig][lzero] = 0L;
        int up = tig == 1 ? ddigits.get(pos) : basee - 1;
        for (int d = 0; d <= up; d++) {
            int new_lz = (lzero == 1 && d == 0) ? 1 : 0;
            int new_pre;
            if (new_lz == 1) {
                new_pre = 0;
            } else {
                if (lzero == 1) {
                    new_pre = d;
                } else {
                    if (d < pre) continue;
                    new_pre = d;
                }
            }
            int new_tig = (tig == 1 && d == up) ? 1 : 0;
            dp[pos][pre][tig][lzero] = (dp[pos][pre][tig][lzero] + dfs(pos + 1, new_pre, new_tig, new_lz)) % MOD;
        }
        return dp[pos][pre][tig][lzero];
    }

    private java.util.List<java.lang.Integer> getBaseBDigits(java.math.BigInteger n, int b) {
        java.util.List<java.lang.Integer> digs = new java.util.ArrayList<java.lang.Integer>();
        if (n.equals(java.math.BigInteger.ZERO)) {
            digs.add(0);
            return digs;
        }
        java.math.BigInteger bb = java.math.BigInteger.valueOf((long) b);
        while (n.compareTo(java.math.BigInteger.ZERO) > 0) {
            digs.add(n.mod(bb).intValue());
            n = n.divide(bb);
        }
        java.util.Collections.reverse(digs);
        return digs;
    }

    private String decrement(String s) {
        return new java.math.BigInteger(s).subtract(java.math.BigInteger.ONE).toString();
    }
}
# @lc code=end

File: 3519.count-numbers-with-non-decreasing-digits.go
#
# @lc app=leetcode id=3519 lang=golang
#
# [3519] Count Numbers with Non-Decreasing Digits
#

# @lc code=start
import (
    "math/big"
)

func countNumbers(l string, r string, b int) int {
    const MOD int64 = 1000000007

    getDigits := func(n *big.Int, base int) []int {
        if n.Cmp(big.NewInt(0)) == 0 {
            return []int{0}
        }
        digits := []int{}
        bb := big.NewInt(int64(base))
        tmp := new(big.Int).Set(n)
        for tmp.Cmp(big.NewInt(0)) > 0 {
            rem := new(big.Int).Mod(tmp, bb)
            digits = append(digits, int(rem.Int64()))
            tmp.Div(tmp, bb)
        }
        // reverse to MSD first
        for i, j := 0, len(digits)-1; i < j; i, j = i+1, j-1 {
            digits[i], digits[j] = digits[j], digits[i]
        }
        return digits
    }

    countLeq := func(s string, base int) int {
        N := new(big.Int)
        N.SetString(s, 10)
        D := getDigits(N, base)
        L := len(D)
        memo := make([][][]int64, L+1)
        for i := 0; i <= L; i++ {
            memo[i] = make([][]int64, 2)
            for j := 0; j < 2; j++ {
                memo[i][j] = make([]int64, base)
                for k := 0; k < base; k++ {
                    memo[i][j][k] = -1
                }
            }
        }
        var dfs func(pos, tight, prev int) int64
        dfs = func(pos, tight1, prev int) int64 {
            if pos == L {
                if prev == 0 {
                    return 0
                }
                return 1
            }
            if memo[pos][tight1][prev] != -1 {
                return memo[pos][tight1][prev]
            }
            ans := int64(0)
            up := base - 1
            if tight1 == 1 {
                up = D[pos]
            }
            for cur := 0; cur <= up; cur++ {
                newTight := 0
                if tight1 == 1 && cur == up {
                    newTight = 1
                }
                newPrev := prev
                cont := true
                if prev == 0 {
                    if cur != 0 {
                        newPrev = cur
                    }
                } else {
                    if cur < prev {
                        cont = false
                    } else {
                        newPrev = cur
                    }
                }
                if cont {
                    ans = (ans + dfs(pos+1, newTight, newPrev)) % MOD
                }
            }
            memo[pos][tight1][prev] = ans
            return ans
        }
        return int(dfs(0, 1, 0))
    }

    countR := countLeq(r, b)

    nl := new(big.Int)
    nl.SetString(l, 10)
    nl.Sub(nl, big.NewInt(1))
    l1 := nl.String()
    countL := countLeq(l1, b)

    res := (int64(countR) - int64(countL) + MOD) % MOD
    return int(res)
}
# @lc code=end